File: core/schema.py
"""Sign Definition Schema — every ASL sign declared as data.

A Sign is a frozen dataclass describing the five linguistic parameters (handshape per hand,
location, movement, palm orientation, non-manual markers), each carrying a `required` flag and
its own confidence threshold. The Phase 3 verifier reads a Sign + a RollingBuffer and gates the
overall pass on EACH required parameter individually — never an average. That gating is what makes
the single-frame COFFEE bug impossible to reproduce.

Structural guard against the single-frame bug (enforced in Sign.__post_init__):
movement is required IF AND ONLY IF a real movement kind is declared. You cannot construct a sign
that declares a movement but marks it not-required, or marks movement required but declares NONE.

All spatial thresholds are RATIOS of shoulder width (never raw pixels) so they hold regardless
of camera distance.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import Optional, Tuple

# Roles — which physical hand a requirement refers to. The verifier maps these to detected
# Left/Right hands; the schema stays handedness-agnostic.
DOMINANT = "dominant"
NONDOMINANT = "nondominant"


# --------------------------------------------------------------------------- handshape
@dataclass(frozen=True)
class HandShapeReq:
    """Required handshape for one hand, matched by core.handshape predicates."""

    kind: str                       # "fist" / "s" / "a" / "index" / "open" / "claw"
    required: bool = True
    min_confidence: float = 0.6


# --------------------------------------------------------------------------- location
class Anchor(str, Enum):
    OTHER_HAND = "other_hand"        # position relative to the other hand's center (two-handed)
    NEUTRAL_SPACE = "neutral_space"  # anywhere in the signing space in front of the torso (loose)
    CHEST = "chest"                  # specifically the center of the chest (below the shoulders)
    CHIN = "chin"                    # the hand must REACH chin height (above shoulders) in the window


@dataclass(frozen=True)
class LocationReq:
    """Where the acting hand must be, normalized to shoulder width."""

    anchor: Anchor = Anchor.OTHER_HAND
    acting_hand: str = DOMINANT
    max_dist_ratio: float = 1.0
    min_dist_ratio: float = 0.0
    vertical: Optional[str] = None           # "above" | "below" | None
    required: bool = True
    min_confidence: float = 0.6


# --------------------------------------------------------------------------- movement
class MovementKind(str, Enum):
    NONE = "none"
    LINEAR = "linear"
    CIRCULAR = "circular"
    REPEATED = "repeated"
    CONVERGE = "converge"            # two hands closing toward each other (e.g. PAIN)


@dataclass(frozen=True)
class MovementReq:
    """How the acting hand must move over the rolling window."""

    kind: MovementKind = MovementKind.NONE
    actor: str = DOMINANT
    pivot: str = NONDOMINANT

    # circular
    min_total_rotation_deg: float = 300.0
    radius_tolerance_ratio: float = 0.4

    # linear
    direction: Optional[Tuple[float, float]] = None
    min_displacement_ratio: float = 0.3

    # repeated
    min_cycles: int = 2

    # converge (PAIN): minimum shrinkage of inter-hand gap, in shoulder-widths
    min_approach_ratio: float = 0.15

    # shared
    min_duration_s: float = 0.6
    required: bool = True
    min_confidence: float = 0.6


# --------------------------------------------------------------------------- orientation
class PalmFacing(str, Enum):
    IN = "in"
    OUT = "out"
    UP = "up"
    DOWN = "down"
    LEFT = "left"
    RIGHT = "right"


@dataclass(frozen=True)
class OrientationReq:
    """Palm-facing requirement for one hand. Off by default in v1."""

    hand: str = DOMINANT
    facing: PalmFacing = PalmFacing.DOWN
    required: bool = False
    min_confidence: float = 0.5


# --------------------------------------------------------------------------- sign
@dataclass(frozen=True)
class Sign:
    """A complete declarative description of one ASL sign."""

    name: str
    dominant: HandShapeReq
    location: LocationReq
    movement: MovementReq
    nondominant: Optional[HandShapeReq] = None
    orientation: Optional[OrientationReq] = None
    nmm: None = None
    two_handed: bool = True

    def __post_init__(self):
        has_motion = self.movement.kind != MovementKind.NONE
        if has_motion and not self.movement.required:
            raise ValueError(
                f"Sign '{self.name}': declares movement kind={self.movement.kind.value} but "
                f"movement.required=False. A declared movement must be enforced."
            )
        if self.movement.required and not has_motion:
            raise ValueError(
                f"Sign '{self.name}': movement.required=True but kind=NONE — nothing to verify."
            )
        if self.two_handed and self.nondominant is None:
            raise ValueError(
                f"Sign '{self.name}': two_handed=True but no nondominant handshape was given."
            )

    def required_parameters(self) -> list[str]:
        params: list[str] = []
        if self.dominant.required:
            params.append("handshape_dominant")
        if self.nondominant is not None and self.nondominant.required:
            params.append("handshape_nondominant")
        if self.location.required:
            params.append("location")
        if self.movement.required:
            params.append("movement")
        if self.orientation is not None and self.orientation.required:
            params.append("orientation")
        return params

File: web/src/engine/schema.ts
export const DOMINANT = 'dominant' as const;
export const NONDOMINANT = 'nondominant' as const;
export type Role = typeof DOMINANT | typeof NONDOMINANT;

export interface HandShapeReq {
  kind: string;
  required: boolean;
  minConfidence: number;
}

export enum Anchor {
  OTHER_HAND = 'other_hand',
  NEUTRAL_SPACE = 'neutral_space',
  CHEST = 'chest',
  CHIN = 'chin',
  FOREHEAD = 'forehead',
  BELLY = 'belly',
  SHOULDER = 'shoulder',
}

export interface LocationReq {
  anchor: Anchor;
  actingHand: Role;
  maxDistRatio: number;
  minDistRatio: number;
  vertical?: 'above' | 'below' | null;
  below?: 'mouth' | null;
  useClosestApproach: boolean;
  required: boolean;
  minConfidence: number;
}

export enum MovementKind {
  NONE = 'none',
  LINEAR = 'linear',
  CIRCULAR = 'circular',
  REPEATED = 'repeated',
  CONVERGE = 'converge',
}

export interface MovementReq {
  kind: MovementKind;
  actor: Role;
  pivot: Role;
  minTotalRotationDeg: number;
  radiusToleranceRatio: number;
  direction?: [number, number] | null;
  minDisplacementRatio: number;
  minCycles: number;
  minAmplitudeRatio: number;
  minApproachRatio: number;
  minDurationS: number;
  required: boolean;
  minConfidence: number;
}

export enum PalmFacing {
  IN = 'in',
  OUT = 'out',
  UP = 'up',
  DOWN = 'down',
  LEFT = 'left',
  RIGHT = 'right',
}

export interface OrientationReq {
  hand: Role;
  facing: PalmFacing;
  required: boolean;
  minConfidence: number;
}

export interface Sign {
  name: string;
  dominant: HandShapeReq;
  location: LocationReq;
  movement: MovementReq;
  nondominant?: HandShapeReq;
  orientation?: OrientationReq;
  twoHanded: boolean;
}

export function createSign(opts: {
  name: string;
  dominant: Partial<HandShapeReq> & { kind: string };
  location: Partial<LocationReq>;
  movement?: Partial<MovementReq>;
  nondominant?: Partial<HandShapeReq> & { kind: string };
  orientation?: Partial<OrientationReq>;
  twoHanded?: boolean;
}): Sign {
  const movement: MovementReq = {
    kind: MovementKind.NONE,
    actor: DOMINANT,
    pivot: NONDOMINANT,
    minTotalRotationDeg: 300,
    radiusToleranceRatio: 0.4,
    direction: null,
    minDisplacementRatio: 0.3,
    minCycles: 2,
    minAmplitudeRatio: 0.05,
    minApproachRatio: 0.15,
    minDurationS: 0.6,
    required: true,
    minConfidence: 0.6,
    ...opts.movement,
  };

  const hasMotion = movement.kind !== MovementKind.NONE;
  if (hasMotion && !movement.required) {
    throw new Error(
      `Sign '${opts.name}': declares movement kind=${movement.kind} but required=false.`
    );
  }
  if (movement.required && !hasMotion) {
    throw new Error(
      `Sign '${opts.name}': movement.required=true but kind=NONE.`
    );
  }

  const twoHanded = opts.twoHanded ?? true;
  if (twoHanded && !opts.nondominant) {
    throw new Error(
      `Sign '${opts.name}': twoHanded=true but no nondominant handshape given.`
    );
  }

  const sign: Sign = {
    name: opts.name,
    twoHanded,
    dominant: { required: true, minConfidence: 0.6, ...opts.dominant },
    location: {
      anchor: Anchor.OTHER_HAND,
      actingHand: DOMINANT,
      maxDistRatio: 1.0,
      minDistRatio: 0.0,
      vertical: null,
      below: null,
      useClosestApproach: false,
      required: true,
      minConfidence: 0.6,
      ...opts.location,
    },
    movement,
  };

  if (opts.nondominant) {
    sign.nondominant = { required: true, minConfidence: 0.6, ...opts.nondominant };
  }
  if (opts.orientation) {
    sign.orientation = {
      hand: DOMINANT,
      facing: PalmFacing.DOWN,
      required: false,
      minConfidence: 0.5,
      ...opts.orientation,
    };
  }

  return sign;
}
